File: Python/1541_잃어버린 괄호.py
import sys

if __name__ == '__main__':
	formula = sys.stdin.readline()
	
	str_num = ''
	oper = []
	num = []
	for f in formula:
		if f == '+' or f == '-': # 연산자 만났을 때
			oper.append(f) # 연산자 저장
			num.append(int(str_num)) # 이전까지 합쳐진 숫자 형태의 문자 숫자로 변환하여 저장
			str_num = '' # 숫자 저장할 문자열 비우기
		else: # 숫자 만났을 때
			str_num += f # 숫자 역시 문자로 받았으므로 문자 concat하듯 합치기
	num.append(int(str_num)) # 마지막 숫자 넣어주기 위함
	
	idx = 0 # 식 계산을 위한 인덱스
	result = num[0] # 가장 앞에 위치한 숫자 초기 값으로 설정
	while idx < len(oper): # 인덱스가 범위를 넘어가기 전까지 루프 수행
		if oper[idx] == '+': # '+' 만났을 때, (현재 위치 앞쪽에 '-'가 없고, '+'를 만난 경우를 의미)
			result += num[idx+1] # 결과 값에 숫자 더해주기(인덱스 주의! 연산자가 숫자보다 1개 적을 것)
			idx += 1 # 인덱스 증가
			continue
		
		# '-' 만난 경우
		sub = num[idx+1] # 부분 합
		tmp_idx = idx + 1 # 부분 합을 위한 인덱스
		while tmp_idx < len(oper) and oper[tmp_idx] == '+': # '-'만나거나 더 이상 계산할 숫자가 없을 때까지(식의 끝)
			sub += num[tmp_idx+1] # 부분 합 구하기
			tmp_idx += 1
		result -= sub # 부분합 모두 구한 후 전체 결과에서 부분 합 빼주고
		idx = tmp_idx # 인덱스 갱신(부분 합 구한 부분 건너 뛰고 다음부터 구해주어야 하기 때문)
		
	print(result)

File: C++/1541_잃어버린괄호.cpp
#include <iostream>
#include <cstring>
#include <vector>
#include <cmath>
using namespace std;

int main() {
	char formula[51];
	cin >> formula;
	
	vector<char> oper;
	vector<int> num;
	
	int len = strlen(formula);
	
	int digit = 4, n = 0; // 최대 5자리를 안 넘기 때문에 10^4부터 10^0까지 주어진 숫자와 곱해주는 과정 필요(일단 모든 수를 5자리라고 가정)
	for(int i = 0; i < len; i++){
		if(formula[i] == '+' || formula[i] == '-'){ // + 또는 -를 만나면
			num.push_back(n/(int)pow(10,digit+1)); // 지금까지 만난 수 계산하여 숫자를 담는 벡터에 넣기(5자리가 아니라면 digit이 0이 안되었을 것.. 그러므로 10^digit을 나눠줘서 주어진 수의 자릿수에 맞게..(불필요한 0 없애기))
			oper.push_back(formula[i]); // 연산자를 넣는 벡터에 연산자(+,-) 넣기
			digit = 4; // 자릿수 초기화
			n = 0; // 숫자를 계산할 값 역시 초기화
			continue;
		}
		
		n += (int)(formula[i]-'0')*(int)pow(10,digit); // char로 받았기 때문에 '0'를 빼주는 과정 필요 / 현재 숫자를 구하기 위한 과정
		digit--; // 한자리 수 감소시켜 한자리에 숫자 하나씩 넣도록
	}
	num.push_back(n/(int)pow(10,digit+1)); // 여기까지 수식 처리... 사실 나눠주는 형식 말고 1~10^4까지 곱해줘도 되는데.. 복잡하게 푼 것 같음ㅠ
	
	int result = num[0]; // 식의 첫 번째 숫자가 초기값이 됨
	for(int i = 0; i < oper.size(); i++){
		if(oper[i] == '+'){ // 이 '+'는 '-'가 한 번도 나오지 않을 때를 대비하여 넣은 것
			result += num[i+1]; // 결과에 더해주기
		}
		else{ // '-' 만났을 때
			int tmp = num[i+1], j = i+1;
			while(j < oper.size() && oper[j] == '+'){ // 다음 '-' 만나거나 마지막 숫자 만날 때까지 부분 합 구해주기
				tmp += num[j+1]; // 부분합 구하기
				j++;
			}
			i = j-1; // 계산한 값들 건너 뛰도록 i를 j-1로 갱신(for문 안의 식으로 인해 i가 1증가할 것이기 때문에 j가 아닌 j-1로 갱신)
			result -= tmp; // 부분 합을 전체 결과에서 빼줌
		}
	}
	
	cout << result;

	return 0;
}
